files/build_templates/docker_image_ctl.j2

```diff
@@ -2,19 +2,51 @@
 
 function getMountPoint()
 {
-    echo $1 |  python -c "import sys, json, os; mnts = [x for x in json.load(sys.stdin)[0]['Mounts'] if x['Destination'] == '/usr/share/sonic/hwsku']; print '' if len(mnts) == 0 else os.path.basename(mnts[0]['Source'])" 2>/dev/null
+    echo $1 | python -c "import sys, json, os; mnts = [x for x in json.load(sys.stdin)[0]['Mounts'] if x['Destination'] == '/usr/share/sonic/hwsku']; print '' if len(mnts) == 0 else os.path.basename(mnts[0]['Source'])" 2>/dev/null
+}
+
+function getRebootType()
+{
+    local TYPE
+    case "$(cat /proc/cmdline)" in
+    *fast-reboot*)
+        TYPE='fast'
+        ;;
+    *warm-reboot*)
+        TYPE='warm'
+        ;;
+    *)
+        TYPE="normal"
+    esac
+    echo $TYPE
 }
 
 function postStartAction()
 {
+    REBOOT_TYPE=`getRebootType`
 {%- if docker_container_name == "database" %}
     until [[ $(/usr/bin/docker exec database redis-cli ping | grep -c PONG) -gt 0 ]]; do
       sleep 1;
     done
+    if [[ "$REBOOT_TYPE" == "warm" && -d /host/warmboot ]]; then
+        WARM_DIR=/host/warmboot
+        function redisLoadAndDelete()
+        {
+            FILENAME="$1"
+            test -e $FILENAME && redis-load -s /var/run/redis/redis.sock -e EMPTY $FILENAME && rm $FILENAME
```

Comment:
A few issues from test:
- rm always fail in this function. -f -rf doesn't help. You might have to remove rm from this action.
- "-s /var/run/redis/redis.sock" cause import to fail always. Removing this option works better.
- import fails randomly. I am stilling looking for a way to make it working reliably. This service is crucial that it has to be reliable.

Reply:
Just a thought:

Maybe we should catch these db restore failures and in case of failure, clear the database and continue with a regular boot up?

Reply:
- rm fixed
- redis-load fixed. if any more failure case, let me know
- I cannot agree to make it retry blindly. I make it exit immediately and we should fix if there is error in normal case.

---
In reply to: [229778573](https://github.com/Azure/sonic-buildimage/pull/2216#discussion_r229778573) [](ancestors = 229778573)

Reply:
I make it exit immediately and we should fix if there is error in normal case.


---
In reply to: [229780996](https://github.com/Azure/sonic-buildimage/pull/2216#discussion_r229780996) [](ancestors = 229780996)

Reply:
My concern is that if we fail database service in product, the device will be in failed state but ASIC is still forwarding. I am not sure if this is better than coming up with cold start and suffer a short IO disruption?